src/template.c

```diff
@@ -5958,22 +5958,23 @@ void TemplateInstance::semantic(Scope *sc, Expressions *fargs)
         this->tnext = inst->tnext;
         inst->tnext = this;
 
-        // If the first instantiation was in speculative context, but this is not:
-        if (tinst && !inst->tinst && !inst->minst)
+        if (minst && minst->isRoot() && !(inst->minst && inst->minst->isRoot()))
         {
-            // Reconnect the chain if this instantiation is not in speculative context.
+            // Swap the minst and tinst between inst and this.
+            Module *mi = minst;
             TemplateInstance *ti = tinst;
-            while (ti && ti != inst)
-                ti = ti->tinst;
-            if (ti != inst)    // Bugzilla 13379: Prevent circular chain
-                inst->tinst = tinst;
-        }
+            minst = inst->minst;
+            tinst = inst->tinst;
+            inst->minst = mi;
+            inst->tinst = ti;
```

Comment:
> It's still confusing that the scopes of this and inst won't match their tinst and minst variables.

The scope of instantiated position, it's the `Scope*sc` parameter of `TemplateInstance::semantic()` function, is not saved anywhere. Therefore the mismatch of tinst/minst won't leak out of the function. And in `TemplateInstance::semantic2()` and `semantic3()`, the given `Scope *sc` parameter is just ignored.

> Will the fact that a parent was a speculative template instance affect the semantic of members?

Never.